Allow `rules!` to have a value

### Current behavior:

```rust
#![feature(label_break_value, proc_macro_hygiene)]

use proc_macro_rules::rules;
use proc_macro2::TokenStream;

fn main() {
    let tokens = TokenStream::new();

    let value = rules!(tokens => {
        ($finish:ident ($($found:ident)*) # [ $($inner:tt)* ] $($rest:tt)*) => {
            0
        }
        (foo $($bar:expr)?) => {
            1
        }
    });

    println!("{}", value);
}
```

```console
error[E0658]: attributes on expressions are experimental. (see issue #15701)
  --> src/main.rs:9:17
   |
9  |       let value = rules!(tokens => {
   |  _________________^
10 | |         ($finish:ident ($($found:ident)*) # [ $($inner:tt)* ] $($rest:tt)*) => {
11 | |             0
12 | |         }
...  |
15 | |         }
16 | |     });
   | |______^
   |
   = help: add #![feature(stmt_expr_attributes)] to the crate attributes to enable

error[E0277]: `()` doesn't implement `std::fmt::Display`
  --> src/main.rs:18:20
   |
18 |     println!("{}", value);
   |                    ^^^^^ `()` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `()`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required by `std::fmt::Display::fmt`

error: aborting due to 2 previous errors
```

### Expected:

`value` should be 0 or 1 depending on which rule matched.

--- a/README.md
+++ b/README.md
@@ -24,11 +24,10 @@ rules!(tokens => {
 
 ## Using proc-macro-rules
 
-Add `proc-macro-rules = "0.1.1"` to your Cargo.toml. You'll need the following feature flags:
+Add `proc-macro-rules = "0.1.1"` to your Cargo.toml. You'll need the following feature flag:
 
 ```rust
 #![feature(proc_macro_hygiene)]
-#![feature(label_break_value)]
 ```
 
 Import the `rules` macro with `use proc_macro_rules::rules`, then use with `rules!(tokens => { branches });` where `tokens` is an expression which evaluates to a `TokenStream` (such as the argument in the definition of a procedural macro).

--- a/macros/src/expand.rs
+++ b/macros/src/expand.rs
@@ -8,16 +8,16 @@ use syn::{Ident, Token};
 crate fn expand_rules(rules: Rules) -> TokenStream2 {
     let clause = rules.clause;
     let rules = rules.rules.into_iter().map(|r| expand_rule(r));
-    quote! {
-        // This is needed because the user can add return statements which make
-        // later code unreachable.
-        #[allow(unreachable_code)]
-        'outer: {
-            let tts = &#clause;
-            #(#rules)*
+    quote!({
+        let tts = &#clause;
+        #(
+            if let Some(value) = #rules {
+                value
+            } else
+        )* {
             panic!("No rule matched input");
         }
-    }
+    })
 }
 
 fn expand_rule(rule: Rule) -> TokenStream2 {
@@ -33,38 +33,43 @@ fn expand_rule(rule: Rule) -> TokenStream2 {
     let builder_name = &rule.name;
     let matches = matches(&builder_name, &variables);
 
-    quote! {
-        {
-            #matches
+    quote!({
+        #matches
 
-            impl syn::parse::Parse for Matches {
-                fn parse(ps: syn::parse::ParseStream) -> syn::parse::Result<Matches> {
-                    let mut ms: proc_macro_rules::MatchSet<#builder_name> = proc_macro_rules::MatchSet::new(ps.fork());
-                    // parse the whole initial ParseStream to avoid 'unexpected token' errors
-                    let _: Result<proc_macro2::TokenStream, _> = ps.parse();
+        impl syn::parse::Parse for Matches {
+            fn parse(ps: syn::parse::ParseStream) -> syn::parse::Result<Matches> {
+                let mut ms: proc_macro_rules::MatchSet<#builder_name> = proc_macro_rules::MatchSet::new(ps.fork());
+                // parse the whole initial ParseStream to avoid 'unexpected token' errors
+                let _: Result<proc_macro2::TokenStream, _> = ps.parse();
 
-                    #rule
+                #rule
 
-                    let result = ms.finalise()?;
-                    // FIXME(#8) pick best match
-                    result.into_iter().filter_map(|p| if p.input.is_empty() {
-                        Some(p.matches.finalise())
-                    } else {
-                        None
-                    }).next().ok_or_else(|| syn::Error::new(proc_macro2::Span::call_site(), "pattern could not be parsed"))
-                }
-            }
-
-            match syn::parse2(tts.clone()) {
-                Ok(Matches { #(#vars,)* }) => {
-                    #body;
-                    break 'outer;
-                }
-                // It can be useful to debug here.
-                Err(e) => {}
+                let result = ms.finalise()?;
+                // FIXME(#8) pick best match
+                result.into_iter().filter_map(|p| if p.input.is_empty() {
+                    Some(p.matches.finalise())
+                } else {
+                    None
+                }).next().ok_or_else(|| syn::Error::new(proc_macro2::Span::call_site(), "pattern could not be parsed"))
             }
         }
-    }
+
+        match syn::parse2(tts.clone()) {
+            Ok(Matches { #(#vars,)* }) => {
+                let value = { #body };
+
+                // This is needed because the body may have ended in a return
+                // statement which makes the `Some` construction unreachable.
+                #[allow(unreachable_code)]
+                {
+                    Some(value)
+                }
+            }
+
+            // It can be useful to debug here.
+            Err(e) => None,
+        }
+    })
 }
 
 fn var_names(variables: &[MetaVar]) -> Vec<Ident> {

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -42,7 +42,6 @@
 
 
 #![feature(proc_macro_hygiene)]
-#![feature(label_break_value)]
 
 extern crate proc_macro2;
 extern crate proc_macro_rules_macros as macros;

